ogr/mock_core.py

```diff
@@ -0,0 +1,93 @@
+import functools
+import logging
+from unittest.mock import MagicMock
+from enum import IntEnum
+from ogr.abstract import PullRequest, PRComment, PRStatus
+import datetime
+
+logger = logging.getLogger(__name__)
+
+
+class MockType(IntEnum):
+    none = 1
+    read_only = 2
+    full = 3
+
+
+def readonly(_func=None, return_value=None, text="", use_internal_args=False):
+    def decorator_readonly(func):
+        @functools.wraps(func)
+        def readonly_func(self, *args, **kwargs):
+            read_only_suffix = "_read_only"
+            if self.mock_type == MockType.none:
+                return func(self, *args, **kwargs)
+            elif self.mock_type == MockType.read_only:
+                internal_readonly_replace_name = func.__name__ + read_only_suffix
+                logger.warning(f"{func.__name__} with args:{args} kwargs:{kwargs}")
+                if internal_readonly_replace_name in dir(self):
+                    return getattr(self, internal_readonly_replace_name)(*args, **kwargs)
+                else:
+                    if use_internal_args:
+                        return return_value(*args, **kwargs)
+                    else:
+                        return return_value
+            elif self.mock_type == MockType.full:
+                logger.warning(f"{text} {func} with args:{args} kwargs:{kwargs}")
+
+                return MagicMock()
```

Comment:
I hope we can avoid these `mock_type` checks as I wrote in the other comment. (And check only `self.read_only`.)

Reply:
solved